File: 2024/day-2.py
# THIS WAS ORIGINALLY DONE IN RUST AND CONVERTED TO PYTHON AFTER THE FACT

import asyncio
from util import main


def part_one(input: str) -> int:
    reports = input.split("\n")
    safe_count = len(reports)

    for report in reports:
        report = [int(n) for n in report.split()]
        dir = 0
        for i in range(0, len(report) - 1):
            step = report[i + 1] - report[i]

            if dir == 0 and step != 0:
                dir = abs(step) / step
            if step == 0 or dir != abs(step) / step or abs(step) > 3:
                safe_count -= 1
                break

    return safe_count


def part_two(input: str) -> int:
    reports = input.split("\n")
    safe_count = len(reports)
    for report in reports:
        report = [int(n) for n in report.split()]
        slopes = [report[i + 1] - report[i] for i in range(len(report) - 1)]
        sign = (sum(slopes) > 0) - (sum(slopes) < 0)

        def valid(n: int):
            return (n > 0) - (n < 0) == sign and 1 <= abs(n) <= 3

        problem_idx = next((i for i, n in enumerate(slopes) if not valid(n)), None)

        if problem_idx is not None:
            i = problem_idx

            shift_right = slopes.copy()
            right_step = shift_right.pop(i)
            if i < len(shift_right):
                shift_right[i] += right_step

            shift_left = slopes.copy()
            left_step = shift_left.pop(i)
            if i > 0:
                shift_left[i - 1] += left_step

            right_problem = any(not valid(n) for n in shift_right)
            left_problem = any(not valid(n) for n in shift_left)

            if right_problem and left_problem:
                safe_count -= 1

    return safe_count


asyncio.run(main(2024, 2, part_one, part_two))

File: 2024/src/solutions/day2.rs
pub struct Solver(pub String);

impl super::lib::Puzzle<i32> for Solver {
    async fn part_one(&self) -> i32 {
        let mut reports = vec![];

        // Consumes the iterator, returns an (Optional) String
        for line in self.0.split("\n") {
            if line.is_empty() {
                break;
            }
            let mut split = line.split_whitespace();
            reports.push(vec![]);
            while let Some(s) = split.next() {
                let n = s.parse::<i32>().unwrap();
                reports.last_mut().unwrap().push(n);
            }
        }

        let mut safe_count = reports.len() as i32;
        for v in reports {
            let mut dir = 0;
            for i in 0..v.len() - 1 {
                let step = v[i + 1] - v[i];
                // setting the initial direction based on the first step
                if dir == 0 {
                    dir = step.signum();
                }
                if dir != step.signum() || step.abs() > 3 || step.abs() == 0 {
                    safe_count -= 1;
                    break;
                }
            }
        }
        safe_count
    }

    async fn part_two(&self) -> i32 {
        let mut reports = vec![];

        // Consumes the iterator, returns an (Optional) String
        for line in self.0.split("\n") {
            let split = line.split_whitespace();
            if line.is_empty() {
                break;
            }
            reports.push(vec![]);
            for s in split {
                let n = s.parse::<i32>().unwrap();
                reports.last_mut().unwrap().push(n);
            }
        }

        let mut safe_count = reports.len() as i32;
        for v in reports {
            let slopes: Vec<i32> = v.windows(2).map(|n| n[1] - n[0]).collect();
            let sign = slopes.iter().sum::<i32>().signum();
            let valid = |n: i32| -> bool { n.signum() == sign && (1..=3).contains(&n.abs()) };
            let problem_idx = slopes
                .iter()
                .enumerate()
                .find(|&(_, &n)| !valid(n))
                .map(|(i, _)| i);
            if let Some(i) = problem_idx {
                let mut shift_right = slopes.clone();
                let right_step = shift_right.remove(i);
                if i < shift_right.len() {
                    shift_right[i] += right_step;
                }
                let mut shift_left = slopes.clone();
                let left_step = shift_left.remove(i);
                if i > 0 {
                    shift_left[i - 1] += left_step;
                }
                let right_problem = shift_right.iter().find(|&&n| !valid(n));
                let left_problem = shift_left.iter().find(|&&n| !valid(n));
                if right_problem.is_some() && left_problem.is_some() {
                    safe_count -= 1;
                }
            }
        }
        safe_count
    }
}
